flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java

```diff
@@ -415,6 +417,12 @@ private void configureStreamGraphStreaming(final StreamGraph graph) {
         graph.setCheckpointStorage(checkpointStorage);
         graph.setSavepointDirectory(savepointDir);
         graph.setGlobalStreamExchangeMode(deriveGlobalStreamExchangeModeStreaming());
+
+        if (Duration.ZERO.equals(
+                configuration.get(
+                        ExecutionCheckpointingOptions.CHECKPOINTING_INTERVAL_DURING_BACKLOG))) {
+            graph.setTimerServiceProvider(InternalBacklogAwareTimerServiceManagerImpl::create);
+        }
```

Comment:
Before this PR, if it is not set, it will use the `InternalTimeServiceManagerImpl::create` as the timerSerivceProvider, see `org.apache.flink.streaming.api.operators.InternalTimeServiceManagerImpl#705`